CBnT: inject measurements made by BootGuard into event log

SergiiDmytruk opened the pull request:
* PCR-0 and (optionally) PCR-7
* TigerLake and newer
* Commit messages and comments in the code provide more details

Was tested for sanity in QEMU which helped to find bugs, but not on hardware (one attempt resulted in a brick although not necessarily related to the new code which doesn't reset the machine or otherwise prevents boot in any way).

*ref: ncm-1811*

SergiiDmytruk committed:
security/tpm: expose tpmalg_from_vb2_hash()

It is useful for more than event log management.

Upstream-Status: Pending
Change-Id: I779f948115af786ecda873cd01c9ef1506252c7d
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
security/intel/cbnt: move union cbnt_biosacm_policy to header

To make it accessible outside of logging.c file.

Upstream-Status: Pending
Change-Id: I3bd7fb126fb197c88bec933a88ded59232d520a1
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
security/intel/cbnt: synthesize PCR-0 measurement

BootGuard/CbnT can optionally measure data into PCRs before passing
control to BIOS which must be taken into account by coreboot for
Measured Boot to work correctly.  This is done by constructing the data
the same way BootGuard does, hashing it and creating an entry in the
event log without extending the corresponding PCR.

BootGuard measurements must precede all coreboot measurements, so the
new code is hooked to tspi_init_crtm() which is responsible for event
log initialization and performing initial measurements (starting with
measuring FMAP).

This also includes 2 other kinds of entries:
 - creation of Startup Locality event on log initialization when startup
   locality of the TPM is locality 3
 - optionally capping PCRs with a separator after a TPM error (the case
   of capping unsupported but active PCR bank is not implemented as
   coreboot hard-codes PCR it works with)

Upstream-Status: Pending
Change-Id: Ib56cadac85d9ef1d747ecf3cfc2976dc6785262a
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
security/intel/cbnt: synthesize PCR-7 measurement

BootGuard/CbnT can optionally measure data into PCRs before passing
control to BIOS which must be taken into account by coreboot for
Measured Boot to work correctly.  This is done by constructing the data
the same way BootGuard does, hashing it and creating an entry in the
event log without extending the corresponding PCR.

Extending PCR-7 is optional and breaks BitLocker on Windows 10, so may
not be utilized.  However, it doesn't require large amount of code and
might as well be implemented along with PCR-0 since both use the same
data structures.

Upstream-Status: Pending
Change-Id: Ic1bfb016600f9f00b8f2fa1965aedfa99647a8e0
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	uint8_t reserved;
+	uint16_t nem_data_stack;  /* in 4 KiB pages */
+	uint8_t se_element[];
+};
```
coreboot add packed atrribute at the ned of structure definition:

```suggestion
struct bpm_header {
	char structure_id[8];     /* "__ACBP__" */
	uint8_t struct_ver;       /* 0x24 */
	uint8_t hdr_struct_ver;   /* 0x20 */
	uint16_t hdr_size;
	uint16_t key_sig_offset;  /* offset to KeySignature field of Policy Manifest Signature
				     Element (PMSE) */
	uint8_t bpm_revision;
	uint8_t bpm_svn;
	uint8_t acmsvn_auth;
	uint8_t reserved;
	uint16_t nem_data_stack;  /* in 4 KiB pages */
	uint8_t se_element[];
} __packed;
```

Please fix for all other cases

SergiiDmytruk replied:
Moved: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R60

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+/*
+ * Definitions of the structures come from Intel document #575623 CBnT BWG v1.2.5 ("Intel
+ * Converged Boot Guard and Intel Trusted Execution Technology (Intel TXT) BIOS Specification").
+ * It is specific for TigerLake and newer platforms.
```
It is not specific. Structures are backward compatible since introduction of Boot Guard.

Some fields were deprecated, but the yare simply marked as reserved.

SergiiDmytruk replied:
Removed that line: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7L49

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	uint8_t pbet_value;
+	uint32_t flags;        /* see BPM_IBBS_FLAG_* constants for values */
+	uint64_t bar0;
+	uint64_t bar1;
```
These are VTD/IOMMU BARs.
```suggestion
	uint64_t iommu_bar0;
	uint64_t iommu_bar1;
```

SergiiDmytruk replied:
Renamed: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R86

miczyg1 commented on src/soc/intel/common/block/cse/cse.c:
```diff
@@ -933,6 +935,58 @@ void print_me_fw_version(void *unused)
+
+	reply_size = sizeof(rmsg);
+
+	if (heci_send_receive(&msg, sizeof(msg), &rmsg, &reply_size, HECI_MKHI_ADDR)) {
```
This function should check the state of ME, not sure if reading a file is allowed in every state of ME.

SergiiDmytruk replied:
It's no longer there (dropped that commit): https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-83770e2316e240c6b33f5ea53347c30dc46faeadcf27d405f50b38562f61b233L974

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	/* Not running any checks on the ACM like TXT code does, it must be correct if we got
+	   here. */
+	size_t acm_len;
+	struct acm_header_v0 *acm = map_cbfs_file(CONFIG_INTEL_TXT_CBFS_BIOS_ACM, &acm_len);
```
I think we should not introduce yet a new API for CBFS access.

Instead, I would simply reference ACM and manifests from FIT directly: https://www.intel.com/content/dam/develop/external/us/en/documents/firmware-interface-table-bios-specification-r1p2p1.pdf

Because we inject manifests as a last step, they cannot be verified with CBFS_VERIFICATION to avoid chicken and egg problem with metadata hash.

There is another reason to prefer the FIT method. If we introduce top swap, then we will have to change the region where to look for ACMs and manifests probably. With FIT method, that would not be necessary as FIT pointer woudl always point to the ACMs and manifests used in current boot.

SergiiDmytruk replied:
Switched to parsing FIT: https://github.com/Dasharo/coreboot/pull/699/commits/49a4b71ffd36822ff25b22469fcf3d7cd9cbf21d and https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R129.

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+			return CB_ERR;
+		}
+	}
+
```
If we try to be so thorough with checking, then I would suggest using `else if` here and also an `else` with `unrecognized key algorithm` error and CB_ERR return.

SergiiDmytruk replied:
The check for unexpected value is in `skip_bpm_signature_key()` called above, so `key_alg` can't be anything but these two values.  But updated to use `else if` for clarity: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R259

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	/*
+	 * ACM.KeyHash[32]
+	 *
+	 * BWG says to use TXT.PUBLIC.KEY, but TXT SDG (#315168-017.4) says that it
+	 * doesn't contain anything useful on modern platforms and suggests SGX-related
+	 * MSRs (0x20::0x23, to be specific, but SDM vol. 4 has these instead).
+	 */
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH0).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH1).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH2).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH3).raw);
```
Hmm. this all looks weird. Also these SGX registers do not exists on all platforms, especially that SGX has been long gone on client machines (last time it is present on CometLake, should not be available on Tiger Lake), onyl exists on servers IIRC.

I did a quick test on my ADL employee laptop with BTG enabled:

```
[miczyg@3M08 coreboot]$ sudo mmio_dump 0xfed30400 0x20
0x00000000fed30400: 0x9c8f9a87 0x1d3d9ebf 0xd79adc12 0xe634de6d
0x00000000fed30410: 0x643640aa 0x34db39c7 0x0b8f857b 0x3aae33e0
[miczyg@3M08 coreboot]$ sudo rdmsr 0x20
350eec1f9866521
[miczyg@3M08 coreboot]$ sudo rdmsr 0x21
de1f549fc0756be
[miczyg@3M08 coreboot]$ sudo rdmsr 0x22
8e2a605e2b89849d
[miczyg@3M08 coreboot]$ sudo rdmsr 0x23
79bd6994c6a790c1
[miczyg@3M08 coreboot]$ sudo rdmsr 0x8c
rdmsr: CPU 0 cannot read MSR 0x0000008c
[miczyg@3M08 coreboot]$ sudo rdmsr 0x8d
rdmsr: CPU 0 cannot read MSR 0x0000008d
[miczyg@3M08 coreboot]$ sudo rdmsr 0x8e
rdmsr: CPU 0 cannot read MSR 0x0000008e
[miczyg@3M08 coreboot]$ sudo rdmsr 0x8f
rdmsr: CPU 0 cannot read MSR 0x0000008f
```

So the MSRs 0x20:0x23 and TXT.PUBLIC.KEY seem to be different. Now to confirm which method is correct I sha256sum'ed the RSA key in ACM header:

```
dd if=ADL_BIOSAC_v1_18_13_20221028_REL_NT_O1.PW_signed_256K.bin skip=128 count=1 bs=384 iflag=skip_bytes |sha256sum
1+0 records in
1+0 records out
384 bytes copied, 5.0509e-05 s, 7.6 MB/s
216586f9c1ee5003be5607fc49f5e10d9d84892b5e602a8ec190a7c69469bd79  -
```

The hash obtained from the ACM header matches what is in the MSR 0x20:0x24 (of course in different endianess).

tl;dr

1. Can't use SGX MSRs on TGL+. Also these MSRs probably contain the enclave key, not ACM key.
2. Can't use TXT.PUBLIC.KEY as BWG suggests.
3. Should use MSRs 0x20:0x23 as BWG suggests.

SergiiDmytruk replied:
Thanks, updated: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R482

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH2).raw);
+	(void)obuf_write_le64(ob, rdmsr(MSR_IA32_SGXLEPUBKEYHASH3).raw);
+
+	/* BP.KEY[32] (rather its SHA256 hash) */
```
Hash algorithm of this file is microachitecture dependent. TGL and onward already provision ME with SHA384 of the key, so the hash read from ME will be SHA384.

```suggestion
	/* BP.KEY[32] (rather its hash) */
```

SergiiDmytruk replied:
Updated: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R376

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	/* BP.KEY[32] (rather its SHA256 hash) */
+	uint8_t bp_key[32];
+	size_t sz = sizeof(bp_key);
+	if (!read_cse_file(OEM_CRED_FILE, bp_key, &sz, 0, MKHI_READ_FILE_FLAG_DEFAULT)) {
```
If ME is not available this command may fail. As a fallback we may read out that key from Key Manifest and hash it with proper algorithm (TGL and onwards SHA384, prior to TGL SHA256). It is better to avoid involving ME if not necessary. We have poor control over it, so it would be best to utilize what we built ourselves (i.e the manfiests).

SergiiDmytruk replied:
Hashing part of Key Manifest now: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R330

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+		return;
+
+	/*
+	 * At this point the code should check whether PCRs 0-7 were capped with a digest of
+	 * value "1" (ASCII? binary? hex?) and if so, add appropriate entries to the event log.
```
The ACM also caps a bank if a bank is enabled bu the appropriate digest of IBB is not in the BPM. E.g. SHA21 enabled bu no SHA1 digest of IBB in BPM, SHA1 bank is capped.

Also the code may simply check TPMFailureAction in BPM IBB Segment Element flags to see what action ACM takes on error. And whether a TPM error occurred, check ACM_POLICY_STATUS bit 15 or the SACM_INFO MSR bit 3.

BWG says that BIOS should log EV_SEPARATOR in the case of capping. It can be found here: https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClient_PFP_r1p05_v23_pub.pdf

```
Per Section 3.3.2 Error Conditions, used to
indicate an error occurred recording the
CRTM, POST BIOS, or Embedded Option
ROMs. For this situation, the digest field
MUST contain the tagged digest of the value
00000001h. The event field is arbitrary to
allow platform manufacturers to include an
indication of what error occurred or other
useful troubleshooting information.
```

So it is a 32bit hex value of "1" that is hashed

SergiiDmytruk replied:
Updated  the comment, didn't introduce querying active banks for handling capping: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R534

miczyg1 replied:
>  didn't introduce querying active banks for handling capping

Leave a TODO then?

SergiiDmytruk replied:
Implemented capping on error and left a TODO about active PCRs: https://github.com/Dasharo/coreboot/compare/4f0d0aa19a293ff4329d846c6f39af07d848a04c..3eb202f68e21e208692f41e80043eddf08d69588#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R520-R580

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	 * Making and hashing PCR-7 data.
+	 *
+	 * Pseudo-code of the data to be measured into PCR-0 for TigerLake and newer (older
+	 * hardware isn't supported):
```
```suggestion
	 * hardware isn't supported yet):
```

But the process is pretty much the same (just slightly different way of obtaining the data).

SergiiDmytruk replied:
Updated: https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R534

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	obuf_init(&data_ob, data, sizeof(data));
+
+	/* ACM_POLICY_STATUS (won't run out of space on this one) */
+	(void)obuf_write_le64(&data_ob, biosacm_sts.raw);
```
Why the `(void)` at the beginning?

SergiiDmytruk replied:
To state that the result of the invocation is being discarded intentionally.

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,543 @@
+	 * the boot anyway as Measured Boot failures generally aren't fatal and incorrect hash
+	 * will be caught later during the boot process.  So not bothering.
+	 */
+
```
I think we are also missing the `10.4.5.3 Startup Locality Event` https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClient_PFP_r1p05_v23_pub.pdf

if TPM is started from locality 3 instead of 0, then such information must be recorded in the event log.

```
The Startup Locality event should be placed in the log before any event which extends PCR[0]. This allows
software which needs to parse the TCG Event Log to initialize its internal PCR[0] state correctly. See Section
3.3.4.1 PCR[0] – SRTM, POST BIOS, and Embedded Drivers.
```

The startup locality is also controlled with a BPM flag and the locality used to perform startup is available in ACM_POLICY_STATUS bit 36

SergiiDmytruk replied:
Iimplemented, should work if whatever processes it allows event's data to be larger than expected: https://github.com/Dasharo/coreboot/pull/699/commits/24f501d37a12c56db9a38321402e32ca43e8a618 and https://github.com/Dasharo/coreboot/compare/ffa516f0285a87e049c9e7ebefc0b23e3d941a77..4b501619643dc15f6eb104ad64919f0fb0b5e347#diff-2191f194b15fa678d26fbe3aa4c3df04b08093f744b0964aca8e2270abcb7910.

miczyg1 commented on src/security/tpm/tspi/log-tpm2.c:
```diff
@@ -22,23 +22,6 @@
-	}
-}
-
 void *tpm2_log_cbmem_init(void)
```
One more thing to that function. The log size should be 64K at minimum per the TCG spec:

```
4. If the ACPI TPM2 table contains the address and size of the Platform Firmware TCG log:
a. The Log Area Minimum Length for the TCG event log MUST be at least 64KB.
```

Here we just allocate 50 entries + size of the log table. Same happens of TPM1.2. Because of that the platform ends up having just a few hundreds of bytes of the log in ACPI TPM table. Not sure if measured boot v2 PRs change that on EDK2 side somehow. Still the EDK2 is not the only payload we use.

SergiiDmytruk replied:
coreboot doesn't actually store measured data in the events, so it's enough for coreboot.  Measured boot v2 stops coreboot from publishing its event log in ACPI so EDK can publish its version which is much larger than what coreboot uses.

miczyg1 replied:
> Measured boot v2 stops coreboot from publishing its event log in ACPI so EDK can publish its version which is much larger than what coreboot uses.

yeah, I know, but when EDK2 is not used, the log ends up being tiny.

SergiiDmytruk replied:
Right, but then nothing except coreboot likely writes to the log.

miczyg1 replied:
@SergiiDmytruk SeaBIOS does IIRC.

SergiiDmytruk replied:
It does (discarding anything that coreboot wrote there), although it respects the length reported in ACPI.  Added the change anyway: https://github.com/Dasharo/coreboot/pull/699/commits/84fb089a040e938ced09b2d74d1b5dd9d17e3348.

miczyg1 replied:
Thank you

SergiiDmytruk committed:
arch/x86/include/arch/fit_table.h: add

Extract the definition from src/cpu/intel/fit/fit_table.c, change first
field to a union and add defines for types.  This provides a basis for
parsing FIT data from coreboot.

Upstream-Status: Pending
Change-Id: If857e21864fd8be2f15a35c154e1f22802fafddd
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
security/tpm/tspi: add a function to log Startup Locality event

The event seems to be specific to TPM2, so not adding to TPM1 log.

Upstream-Status: Pending
Change-Id: I99b9c23a5089b039b01090eec08a808346f7c389
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

miczyg1 commented on src/arch/x86/include/arch/fit_table.h:
```diff
@@ -0,0 +1,82 @@
+{
+	struct fit_entry *fit = *(struct fit_entry **)(uintptr_t)FIT_POINTER_ADDRESS;
+	int entry_count = fit_entry_size(fit) * 16 / sizeof(struct fit_entry);
+
```
Shouldn't we check if the first FIT entry has the proper signature?

SergiiDmytruk replied:
Added: https://github.com/Dasharo/coreboot/compare/4f0d0aa19a293ff4329d846c6f39af07d848a04c..3eb202f68e21e208692f41e80043eddf08d69588#diff-b3cae2b23269e1bd1b562bf387e9bc0017526cd57b2979919b9e36995a9addbc

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,621 @@
+	 *
+	 * BWG says to use TXT.PUBLIC.KEY, but TXT SDG (#315168-017.4) says that it
+	 * doesn't contain anything useful on modern platforms and suggests SGX-related
+	 * MSRs (0x20::0x23, to be specific, but SDM vol. 4 has these instead).
```
Where is that suggestion in SDM? I think this comment should be removed/changed.

SergiiDmytruk replied:
`B.1.21 TXT.PUBLIC.KEY – AC Module Public Key Hash`:

> Note: Modern platforms utilizing SGX technology implement the Public Key Hash register as
part of the CPU core, not as part of PCH and thus TXT.PUBLIC.KEY register has been
deprecated with no underlying HW functionality. Any value it may contain does not have
meaning and must be disregarded.
Actual Public Key Hash is typically available in CPU MSRs 0x20∷0x23.
Platform manufacturers shall consult respective external data sheets to verify the above
numeric values as it may change.

Also mentioned when `EVTYPE_SINIT_PUBKEY_ HASH` is described:
> PUBKEY_HASH is digest of ACM Signer key. It can be retrieved from MSRs 0x20::0x23

miczyg1 replied:
That's a note from MLE SDG. SDM is a different document. But I already see where it came from.

1. MLE SDG and CBNT spec are right when it comes to MSRs 0x20:0x23, which I already proved.
2. We have to remove any reference of SDM here. It seems you have mistaken SGXLEPUBKEYHASH with ACM PUBKEYAHSH MSRs, because both have PUBKEYHASH in its names. But the ACM PUBKEY MSRS are nto SGX-related at all. Just the information about ACM pubkey hash has been moved from PCH MMIO registers to CPU MSRs since the introduction of SGX technology. However, SGX has been removed after Comet Lake (10th Gen) from client devices, but the MSRs 0x20:0x23 are still there, because they have different purpose and are not SGX related.

Regardless of the reason, SDM does not have any information about Boot Guard event logging, nor about MSRS 0x20:0x23. Thus the comment should be rephrased and not mention SDM at all:

```
	 * BTG BWG says to use TXT.PUBLIC.KEY, but TXT SDG (#315168-017.4) and CBnT BWG
	 * says that it doesn't contain anything useful on modern platforms utilizing SGX and
	 * suggests to use MSRs 0x20::0x23.
```

One may also calculate the hash of the key from ACM header itself and compare against TXT.PUBLIC.KEY or the MSRs to be sure. But I think the Skylake+ (6th gen) will have the key in MSRs already.

SergiiDmytruk replied:
> 2\. It seems you have mistaken SGXLEPUBKEYHASH with ACM PUBKEYAHSH MSRs, because both have PUBKEYHASH in its names.

That wasn't a mistake.  I picked it because I didn't find anything that fits the description nor description of MSRs 0x20:0x23 in SDM, and since SGX was mentioned in both places decided to go with it which is what the comment was explaining.  I should have updated it when switching to 0x20:0x23, did it now: https://github.com/Dasharo/coreboot/compare/5820a99a5ee7959893db963edf3b597bc5e63554..84fb089a040e938ced09b2d74d1b5dd9d17e3348#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R494

miczyg1 replied:
> I picked it because I didn't find anything that fits the description nor description of MSRs 0x20:0x23 in SDM

Still don't understand why using SDM here, but never mind.

SergiiDmytruk replied:
> Still don't understand why using SDM here

Because Volume 4 of Intel 64 and IA-32 Architectures Software Developer’s Manual describes MSRs.

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,621 @@
+			   always returns the first one which might be wrong. */
+			printk(BIOS_WARNING, "CBnT: picking SACM from FIT at random!\n");
+		}
+		*size = ((struct acm_header_v0 *)addr)->size * 4;
```
The `acm_header_v0` might need an update. There are newer versions of acm_header. I believe ADL or MTL uses v3 already which has slightly different structure (RSAPubKey/RSASig is bigger, depends on KeySize field and RSAPubExp is omitted, Scratch and User Area also get shifted). SO maybe a union of acm header versions would be needed.

SergiiDmytruk replied:
Key data is at different offsets indeed, thanks.

v3 should cover our needs, it's also the first version with CBnT support.  TGL and ADL use v3, MTL has v4.  The signature part is compatible between v3.0 and v5.4 (v5 seems to have no change for the header compared to v4, just version number change).

code changes: https://github.com/Dasharo/coreboot/compare/4f0d0aa19a293ff4329d846c6f39af07d848a04c..3eb202f68e21e208692f41e80043eddf08d69588#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7R145-R187
declaration: https://github.com/Dasharo/coreboot/compare/4f0d0aa19a293ff4329d846c6f39af07d848a04c..3eb202f68e21e208692f41e80043eddf08d69588#diff-1013ce554074c41208e8c9b5239fbe83423c40bf24d9a07ff5b98440bba4a74c

miczyg1 replied:
Is it good that we assume V3 always? Maybe a union would be better + a check for ACM header version.

SergiiDmytruk replied:
Added a version check in https://github.com/Dasharo/coreboot/compare/5820a99a5ee7959893db963edf3b597bc5e63554..84fb089a040e938ced09b2d74d1b5dd9d17e3348#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7.

I wanted to add a union last time, but it wouldn't do anything useful in this case (v4/v5 adds and modifies some fields at the end), so I didn't even bother committing the header for v4.

SergiiDmytruk reviewed (commented):
Other changes apart from rebasing and review:
 * removed Q35 and CBFS leftovers
 * made more things `const` which wasn't possible due to CBFS
 * added error checks after FIT lookup since `find_in_fit()` is returning `NULL` on error

SergiiDmytruk committed:
security/tpm/tspi: allocate 64 KiB for TCG logs

That's the minimum log size for client platforms per "TCG PC Client
Platform Firmware Profile Specification" [0].

[0]: https://trustedcomputinggroup.org/resource/pc-client-specific-platform-firmware-profile-specification/

Upstream-Status: Pending
Change-Id: Ifed85a91d5e2cf28fcc40262710e0ec688479a62
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,675 @@
+	switch (type) {
+		/* All known versions of the header (up to v5.4) have compatible version and
+		   size fields. */
+		const struct acm_header_v0 *acm;
```
Variable declarations should be on top of the function content block.

SergiiDmytruk replied:
Moved the declaration before the `switch`: https://github.com/Dasharo/coreboot/compare/84fb089a040e938ced09b2d74d1b5dd9d17e3348..4a20083b3990dba0883f92b1f90408bbbb276a61#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7

Top of the function would be farther away from the use and inconsistent with other declarations in this and other functions.

miczyg1 commented on src/security/intel/cbnt/measurement.c:
```diff
@@ -0,0 +1,675 @@
+		}
+
+		acm = (const struct acm_header_v0 *)addr;
+		if (acm->header_version[1] > 5 || acm->header_version[0] > 4) {
```
The important comment wandered to the top of switch statement, but was probably meant to be here:

```
		/* All known versions of the header (up to v5.4) have compatible version and
		   size fields. */
```

SergiiDmytruk replied:
Moved the comment here: https://github.com/Dasharo/coreboot/compare/84fb089a040e938ced09b2d74d1b5dd9d17e3348..4a20083b3990dba0883f92b1f90408bbbb276a61#diff-1725d4cd7234cc446584ac6ebad791ed8c8e640970d4adb2e99e5b5b233030f7

miczyg1 reviewed (dismissed):
Minor code style problems left

macpijan commented:
@miczyg1 I would appreciate dropping the `request change` status if your comments  are addressed.

EDIT:

Ok @mkopec could also do that for these style fixes.
